src/test/java/htsjdk/variant/bcf2/BCF2WriterUnitTest.java

```diff
@@ -0,0 +1,253 @@
+        final File fakeVCFFile = VariantBaseTest.createTempFile("testWriteAndReadVCFBody.", ".bcf");
+
+        Tribble.indexFile(fakeVCFFile).deleteOnExit();
+        metaData = new HashSet<VCFHeaderLine>();
+        additionalColumns = new HashSet<String>();
+        final SAMSequenceDictionary sequenceDict = createArtificialSequenceDictionary();
+        final VCFHeader header = createFakeHeader(metaData, additionalColumns, sequenceDict);
+        try (final VariantContextWriter writer = new VariantContextWriterBuilder()
+                .setOutputFile(fakeVCFFile).setReferenceDictionary(sequenceDict)
+                .setOptions(EnumSet.of(Options.ALLOW_MISSING_FIELDS_IN_HEADER, Options.INDEX_ON_THE_FLY))
+                .build()) {
+            writer.writeHeader(header);
+            writer.add(createVC(header));
+            writer.add(createVC(header));
+        }
+        VariantContextTestProvider.VariantContextContainer container = VariantContextTestProvider
+                .readAllVCs(fakeVCFFile, new BCF2Codec());
+        int counter = 0;
+        final Iterator<VariantContext> it = container.getVCs().iterator();
+        while (it.hasNext()) {
+            it.next();
+            counter++;
+        }
+        Assert.assertEquals(counter, 2);
+    }
+
+    /**
+     * test, using the writer and reader, that we can output and input a BCF body without header
+     */
+    @Test
+    public void testWriteAndReadBCFBody() throws IOException {
+        final File fakeVCFFile = VariantBaseTest.createTempFile("testWriteAndReadVCFBody.", ".bcf");
+        Tribble.indexFile(fakeVCFFile).deleteOnExit();
+        final File fakeVCFBodyFile = VariantBaseTest.createTempFile("testWriteAndReadVCFBody.", ".bcf");
+        Tribble.indexFile(fakeVCFBodyFile).deleteOnExit();
+
+        metaData = new HashSet<VCFHeaderLine>();
+        additionalColumns = new HashSet<String>();
+        final SAMSequenceDictionary sequenceDict = createArtificialSequenceDictionary();
+        final VCFHeader header = createFakeHeader(metaData, additionalColumns, sequenceDict);
+
+        // we write two files, fakeVCFFile with the header and body, and fakeVCFBodyFile with just the body
+        try (final VariantContextWriter fakeVCFFileWriter = new VariantContextWriterBuilder()
+                .setOutputFile(fakeVCFFile).setReferenceDictionary(sequenceDict)
+                .setOptions(EnumSet.of(Options.ALLOW_MISSING_FIELDS_IN_HEADER, Options.INDEX_ON_THE_FLY))
+                .build()) {
+            fakeVCFFileWriter.writeHeader(header); // writes header
+            fakeVCFFileWriter.add(createVC(header));
+            fakeVCFFileWriter.add(createVC(header));
+        }
+
+        try (final VariantContextWriter fakeVCFBodyFileWriter = new VariantContextWriterBuilder()
+                .setOutputFile(fakeVCFBodyFile).setReferenceDictionary(sequenceDict)
+                .setOptions(EnumSet.of(Options.ALLOW_MISSING_FIELDS_IN_HEADER, Options.INDEX_ON_THE_FLY))
```

Comment:
Creating an index on one of these header-less files seems questionable. I wonder if we can/should prevent that, or if we should allow it.